service/dap/server.go

```diff
@@ -686,16 +688,189 @@ func (s *Server) onStackTraceRequest(request *dap.StackTraceRequest) {
+	if !ok {
+		s.sendErrorResponse(request.Request, UnableToListLocals, "Unable to list locals", fmt.Sprintf("unknown frame id %d", request.Arguments.FrameId))
+		return
+	}
+
+	scope := api.EvalScope{GoroutineID: sf.(stackFrame).goroutineID, Frame: sf.(stackFrame).frameIndex}
+	// TODO(polina): Support setting config via launch/attach args
+	cfg := &api.LoadConfig{FollowPointers: true, MaxVariableRecurse: 1, MaxStringLen: 64, MaxArrayValues: 64, MaxStructFields: -1}
+
+	// Retrieve arguments
+	args, err := s.debugger.FunctionArguments(scope, *api.LoadConfigToProc(cfg))
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToListArgs, "Unable to list args", err.Error())
+		return
+	}
+	argScope := api.Variable{Name: "Arguments", Children: args}
+
+	// Retrieve local variables
+	locals, err := s.debugger.LocalVariables(scope, *api.LoadConfigToProc(cfg))
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToListLocals, "Unable to list local vars", err.Error())
+		return
+	}
+	locScope := api.Variable{Name: "Locals", Children: locals}
+
+	// TODO(polina): Annotate shadowed variables
+	// TODO(polina): Retrieve global variables
+
+	scopes := make([]dap.Scope, 0)
+	scopes = append(scopes, dap.Scope{Name: argScope.Name, VariablesReference: s.variableHandles.create(argScope)})
+	scopes = append(scopes, dap.Scope{Name: locScope.Name, VariablesReference: s.variableHandles.create(locScope)})
+
+	response := &dap.ScopesResponse{
+		Response: *newResponse(request.Request),
+		Body:     dap.ScopesResponseBody{Scopes: scopes},
+	}
+	s.send(response)
 }
 
-// onVariablesRequest sends a not-yet-implemented error response.
+// onVariablesRequest handles 'variables' requests.
 // This is a mandatory request to support.
-func (s *Server) onVariablesRequest(request *dap.VariablesRequest) { // TODO V0
-	s.sendNotYetImplementedErrorResponse(request.Request)
+func (s *Server) onVariablesRequest(request *dap.VariablesRequest) {
+	variable, ok := s.variableHandles.get(request.Arguments.VariablesReference)
+	if !ok {
+		s.sendErrorResponse(request.Request, UnableToLookupVariable, "Unable to lookup variable", string(request.Arguments.VariablesReference))
+		return
+	}
+	v := variable.(api.Variable)
+	children := make([]dap.Variable, 0)
+	// TODO(polina): check and handle if variable loaded incompletely
+	// https://github.com/go-delve/delve/blob/master/Documentation/api/ClientHowto.md#looking-into-variables
+
+	switch v.Kind {
+	case reflect.Map:
+		for i := range v.Children {
+			// Process items in pairs: even indices are map keys, odd indices are values.
+			if i%2 == 0 {
+				continue
+			}
+			key, keyref := s.convertVariable(v.Children[i-1])
+			val, valref := s.convertVariable(v.Children[i])
+			// If key or value or both are scalars, we can use
+			// a single variable to represet key:value format.
+			// Otherwise, we must return separate variables for both.
+			if keyref > 0 && valref > 0 { // Both are not scalars
```

Comment:
OK, but please add a comment in handles.go on the create() method that it's guaranteed that 0 is not a valid handle. Right now the constant `startHandle` ensures that, but for a person modifying just `handles.go` it's not clear why this value is special and why 0 wouldn't work just as well.

Reply:
I decided to sleep on this a bit, and the comment still does not make sense to me. The only guarantee that handles map has is that values below startHandle won't be represented because of how things get created. There is nothing special about 0 vs 1 or even 999 when startHandle is 1000. It is the dap client that treats 0 differently from 1. 2, 10, etc. If you send 999 back to the client, it will happily request a look-up for that number, which will come back with an error. But if you send a 0, then it will not even request anything, so the look up will never happen at all, regardless of what handles.go guarantees. So should that really be explained in handles.go? 

What I could do is the following:
(1) Add explicit documentation that handles are guaranteed to be positive
(2) Add explicit documentation on choice of startHandle and that all handles will be higher than that number
(3) Make startHandle an arg of newHandlesMap, so all limitations are enforced by the caller.
What do you think?

Reply:
I prefer not to hold up the review upon this comment. We can discuss it separately.